README.md

eZLeafletLocation Datatype Extension
Version 1.x by eZ Systems AS

Cloned and adapted from [EZGmapLocation Extension](https://github.com/ezsystems/ezgmaplocation-ls-extension)


Version 0.6 developed by [Smile](https://smile.eu)

------------------------------------------------------------------

The LeafletLocation datatype extension provides a handy way to store
latitude/longitude points (as decimal degrees) on an object by using
Leafletjs to identify and mark positions using their address.


Installation
---------------

1.) Upload the ezleafletlocation folder to the extensions folder in your
eZ Publish installation.

2.) Activate the extension from the 'Extensions' portion of the
'Setup' tab in the eZ publish admin interface.
And update the autoload array by clicking "Regenerate autoload arrays for extensions"

**Project configuration editing via BO interface is discouraged :**
- in your "settings/override/site.ini.append.php" add those lines :
```
ActiveExtensions[]=ezleafletlocation #under [ExtensionSettings]
...
TranslationExtensions[]=ezleafletlocation #under [RegionalSettings]
```
- clear caches ("ini" mainly)
- and run ```php bin/php/ezpgenerateautoloads.php```


3.) Apply the ezleafletlocation table to your database with the included sql file:
```extension/ezleafletlocation/sql/mysql/mysql.sql```
Using either phpmyadmin (easiest) or shell/console commands.

Sql files are also provided for postgressql and oracle - refer to the
database documentation on how to execute queries from a command-line clients

4.) Now you can add the ezleafletlocation datatype like any other datatype when editing classes.

5.) If you wish to quickly replace your current eZGmapLocation class and object attributes, you can run this script :
``` extension/ezleafletlocation/bin/replacegmapswithleaflet.php ```

Reversely, if you wish to step back with eZLeafletLocation class and object attributes replacement with eZGmapLocation, run this script :
``` extension/ezleafletlocation/bin/replaceleafletwithgmaps.php ```

**Backup your data before using it**

Use (editing)
---------------
- Type in the address you want to find beneath the map.
- Click 'Find address'
or
Type in longitude / latitude input
- Click on "My current location" to geolocate yourself (https only)
- Click 'Restore' to cancel modifications.

Repeat to change.
Unless you change the marker on the map, the address will be saved as you typed it.
This address is searchable! (see bellow for filtering / sorting on coordinates as well)

![ezleafletdemo](https://j.gifs.com/nrWv35.gif)


Use (fetching)
---------------

For fetching multiple nodes based on location, you can use the included ezlfltLocationFilter.
Example fetches users in a distance of roughly 30-50km from Oslo to Norway. And sorts the
results based on how close the nodes are to the given coordinate.

```smarty
{def $users_close_by = fetch( 'content', 'tree', hash(
                              'parent_node_id', 12,
                              'limit', 3,
                              'sort_by', array( 'distance', true() ),
                              'class_filter_type', 'include',
                              'class_filter_array', array( 'user' ),
                              'extended_attribute_filter', hash( 'id', 'ezlfltLocationFilter', 'params', hash( 'latitude', 59.917,
                                                                                                              'longitude', 10.729,
                                                                                                              'distance', 0.5 ) )
                              ) )}
```

 Note that the distance filter is using a 'bounding box' for sql speed, see classes/ezlfltlocationfilter.php for more info
 and paramerters to be able to get true ('arccosine') or closer to true ('pythagorean') circular distance filter accuracy.
 The sort on the other hand is accurate, so if your main concerne is to show closets node, then the filter is ok by default.
 
 Also see 'arccosine' parameter in combination with 'as_object', false() to be able to get value to easily calculate distance.
 Example use: user x is 2,5 km away from you. Or Oslo, Norway is 416.8km from Stockholm, Sweden.



CHANGELOG.txt
CHANGELOG

Version 1.3
-------------------------------------------------------------------------------
* Conversion of Google Maps API / Google Maps Platform Solution to Open Source Leaflet solution
* Conversion of ezgmapslocation to ezleafletlocation attributes script addition

Version 1.2
-------------------------------------------------------------------------------
* Change use of old templateInit()/ezi18n() api and use class based api's introduced in 4.3 instead

Version 1.1
-------------------------------------------------------------------------------
* Added extension.xml. Replaces loading.php and ezinfo.php starting from ezp4.4
* Added nor-NO translation
* Fixed bug #16160: Call to undefined function ezi18n()
* Implemented enhancement #016711: Making extension aware of the
  order/priority in which they need to be loaded
* added ini file to silence wrong suggestions to alter oracle db in system
  upgrade check page

Version 1.0
-------------------------------------------------------------------------------
* Removed Update Location button and made location update automatically on
  marker change.
* Added restore button to restore values to value it had on page load.
* Clear overlay markers on find address click and restore click
* Make edit ui more intuitive by rearranging buttons and use button class
* Add current location button to get location from GeoLocation api if available

Version 1.0beta1
-------------------------------------------------------------------------------
* code cleanup
* bigger map in edit view and normal view
* button to grab current location (from bcgmapslocation)
* PHPDoc inline documentation added where applicable
* Address is now stored
* Datatype can now be used as a name pattern component (address)
* Datatype is now searchable (by address)
* Datatype is now sortable (by address)
* Store data in external table to be able to filter on it
* Added Extended attribute filter to fetch list / tree by cordinates and distance
* Added support for sorting by distance when using Extended attribute filter
* Added support for arccosine and pythagorean distance algorithm in distance
  calculation in the Extended attribute filter in addition to bounding box
* added sql file for Oracle databases

Version 0.5
-------------------------------------------------------------------------------
* only compatible with eZ Publish 4

Version 0.4
-------------------------------------------------------------------------------
* Resolved a few warnings and error messages.

* Much simpler install - modified the javascript so that no template changes
  are required in the admin design.

* The editing controls will now work properly if there are multiple location
  attributes on a single object.

* In the editing view, the map will now re-center on the clicked point. This
  makes it easier to use the click-point and zoome controls to hone in on a
  specific point.

* The edit map now defaults to a zoom-level-0 view of the globe instead of
  a gray square.

* The edit map is larger, so that the map type and directional controls don't
  overlap.

* A new template has been included, gmap.tpl. This is a multi-purpose template
  that can be included and configured to handle several common use-cases for
  the gmaps datatype. See the doc on the template for specifics.

Version 0.3 - Initial Release
-------------------------------------------------------------------------------


classes/ezlfltlocationfilter.php
<?php
//
// Definition of ezlfltLocationFilter class
//
// Created on: <10-Aug-2009 12:42:08 ar>
//
// ## BEGIN COPYRIGHT, LICENSE AND WARRANTY NOTICE ##
// SOFTWARE NAME: eZ Google Maps Location
// SOFTWARE RELEASE: 1.3.1
// COPYRIGHT NOTICE: Copyright (C) 2009-2010 eZ Systems AS
// SOFTWARE LICENSE: GNU General Public License v2.0
// NOTICE: >
//   This program is free software; you can redistribute it and/or
//   modify it under the terms of version 2.0  of the GNU General
//   Public License as published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of version 2.0 of the GNU General
//   Public License along with this program; if not, write to the Free
//   Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
//   MA 02110-1301, USA.
//
//
// ## END COPYRIGHT, LICENSE AND WARRANTY NOTICE ##
//

class ezlfltLocationFilter
{
    function ezlfltLocationFilter()
    {
    }

    function createSqlParts( $params )
    {
        /*
         * Filtering nodes within a area relative to supplied coordinates.
         * 
         * Params can be either hash or array, in case of array this is the order:
         * param 1: latitude
         * param 2: longitude
         * param 3: distance (Optional radius length, default is 0.5)
         * param 4: arccosine (Optional, calculate accurate distance, default is false())
         * param 5: distance_as_circle (Optional, see 'NOTE on distance accuracy', default is false())
         * param 6: arccosine_distance (Optional, see 'NOTE on arccosine_distance', default is 0.005)
         * 
         * Note: All params needs to be numeric (incl numeric string values, lik '5.5')
         * 
         * Full example for fetching nodes that are within a certain area:
         * 
         * {def $users_in_oslo = fetch( 'content', 'tree', hash(
                                      'parent_node_id', 1503,
                                      'limit', 3,
                                      'sort_by', array( 'distance', true() ),
                                      'class_filter_type', 'include',
                                      'class_filter_array', array( 'user' ),
                                      'extended_attribute_filter', hash( 'id', 'ezlfltlLocationFilter', 'params', hash( 'latitude', 59.917, 'longitude', 10.729, 'distance', 0.35 ) )
                                      ) )}
         *
         * NOTE on distance accuracy:
         * This filter uses a bounding box to make sure the sql is fast, this means distance is actually a square and not circle by default.
         * If you want more accuracy in the filter, use 'distance_as_circle' setting:
         *    false()       : Default, fastest, no additional calculation other then the bounding box (square) distance
         *    'pythagorean' : Use somewhat fast pythagorean calculation, accurate for small distances ( < 60km  )
         *    'arccosine'   : Use full arccosine calculation, in this case provide 'arccosine_distance' as well
         *
         * NOTE on arccosine_distance:
         * A double number, if you have distance in km divide by 6371, if miles divide by 3959
         * For use in addition to 'distance' when 'distance_as_circle' is set to 'arccosine'
         *
         * NOTE on sorting:
         * You can optionally sort on 'distance' when this filter is used.
         *   
         * NOTE on as_object = false:
         * If you fetch content using the 'as_object' parameter set to false(), you'll get additional columns:
         *     latitude
         *     longitude
         *     distance  A simple pythagorean calculated distance, only for use in sorting, as it's not
         *               accurate enough to be used for basis of km and miles.
         *     arccosine An accurate distance calculation, multiply by 6371 for km and 3959 for miles
         *               Is not included by default, include by setting 'arccosine' to true()
         *                 
         *        
         * CREDITS:
         * arccosine SQL based on "Beginning Google Maps Mashups with Mapplets, KML & GeoRSS: From Novice
         * to Professional (Expert's Voice in Web Development)" by Sterling Udell, page 257
         *
         * Also found as "Spherical Law of Cosines" on http://www.movable-type.co.uk/scripts/latlong.html
         * and http://www.movable-type.co.uk/scripts/gis-faq-5.1.html in JavaScript and Excel code
         * 
         */


        $latitude  = isset( $params['latitude'] ) && is_numeric( $params['latitude'] ) ? $params['latitude'] : ( isset( $params[0] ) && is_numeric( $params[0] ) ? $params[0] : 59.917 );
        $longitude = isset( $params['longitude'] ) && is_numeric( $params['longitude'] ) ? $params['longitude'] : ( isset( $params[1] ) && is_numeric( $params[1] ) ? $params[1] : 10.729 );
        $distance  = isset( $params['distance'] ) && is_numeric( $params['distance'] ) ? $params['distance'] : ( isset( $params[2] ) && is_numeric( $params[2] ) ? $params[2] : 0.5 );
        $arccosine = isset( $params['arccosine'] ) ? $params['arccosine'] : ( isset( $params[3] ) ? $params[3] : false );
        $asCircle  = isset( $params['distance_as_circle'] ) ? $params['distance_as_circle'] : ( isset( $params[4] ) ? $params[4] : false );
        $aDistance = isset( $params['arccosine_distance'] ) && is_numeric( $params['arccosine_distance'] ) ? $params['arccosine_distance'] : ( isset( $params[5] ) && is_numeric( $params[5] ) ? $params[5] : 0.005 );

        $minLatitude = $latitude - $distance;
        $maxLatitude = $latitude + $distance;
        $minLongitude = $longitude - $distance;
        $maxLongitude = $longitude + $distance;

        $sqlSelect = ",
            ezleafletlocation.latitude AS latitude,
            ezleafletlocation.longitude AS longitude,
            SQRT(POW(ezleafletlocation.latitude - $latitude,2) + POW( ezleafletlocation.longitude - $longitude,2)) AS distance";

        if ( $arccosine )
        {
            $sqlSelect .= ",
                acos( cos(radians( $latitude )) * cos(radians( ezleafletlocation.latitude )) *
                     cos(radians( ezleafletlocation.longitude  -  $longitude )) +
                     sin(radians( $latitude )) * sin( radians( ezleafletlocation.latitude ))) AS arccosine_distance";
        }

        $sqlWhere = "ezleafletlocation.latitude >= $minLatitude AND ezleafletlocation.latitude <= $maxLatitude AND
                     ezleafletlocation.longitude >= $minLongitude AND ezleafletlocation.longitude <= $maxLongitude AND";

        if ( $asCircle === 'pythagorean' )
        {
            $sqlWhere .= "
                     SQRT(POW(ezleafletlocation.latitude - $latitude,2) + POW(ezleafletlocation.longitude - $longitude,2)) <= $distance AND";
        }
        else if ( $asCircle === 'arccosine' )
        {
            $sqlWhere .= "
                     acos( cos(radians( $latitude )) * cos(radians( ezleafletlocation.latitude )) *
                     cos(radians( ezleafletlocation.longitude  -  $longitude )) +
                     sin(radians( $latitude )) * sin( radians( ezleafletlocation.latitude ))) <= $aDistance AND";
        }

        $sqlWhere .= "
                     ezcontentobject_attribute.id = ezleafletlocation.contentobject_attribute_id AND
                     ezcontentobject_attribute.version = ezleafletlocation.contentobject_version AND
                     ezcontentobject.id = ezcontentobject_attribute.contentobject_id AND
                     ezcontentobject.current_version = ezcontentobject_attribute.version AND";

        return array('tables' => ', ezcontentobject_attribute, ezleafletlocation', 'joins' => $sqlWhere, 'columns' => $sqlSelect );
    }
}
?>